### DISCUSSION TITLE
I evaluated the performance of the model by backtesting it on the 2022-23 season. (so far)

### DISCUSSION THREAD

**Discussion by ghost** [kyleskom/NBA-Machine-Learning-Sports-Betting#69] [Show and tell]
I backtested the "XGBoost_74.9%_ML-2.json" model on the 2022-23 season.

I downloaded data from teams using the `Get_Data` module, and then created a new function to acquire all historical odds from FanDuel using the `SbrOddsProvider` library. Next, I executed the `Create_Games` module to create the 2022-23 dataset.

The 2022-23 dataset now contains two additional columns: `home_team_odds` and `away_team_odds`. These are necessary for the expected value calculation.

Once all steps were completed, I ran the `Test_Current_Season` script. Rather than returning only the winner prediction, I made a few modifications to the code so that it now returns the expected value, the probability, and the winner prediction. It only returns the team with the highest expected value.

For a $1 dollar per bet, the graph starts one month after the season start.

![image](https://user-images.githubusercontent.com/43114210/215239389-e3437d2d-f258-4a24-bab7-32359d3615b6.png)

I utilized WPS Spreadsheets, and encountered a few difficulties while attempting to upload to Google Sheets and online Excel. I will provide the link to the .xlsx file, but here is a picture that displays all the information on the spreadsheet:

![Current-Season-Test](https://user-images.githubusercontent.com/43114210/215239293-5d9ef0d2-260d-46c2-86be-ee45d6a3e19b.png)

Link to .xlsx file: https://mega.nz/file/YNgzxBwa#fo3Sp10MeefLl22z6qNAU5e2s7I4Q9MaOPd9s2ziAn8

**Comment by kyleskom** [comment]
Thanks for sharing this. I think seeing stats like this a helpful and allow us to visualize different strategies based on EV value or win probabilities and make bets accord to certain thresholds. One question, for the second column prediction they are all 1, is that expected?

**Reply by ghost** [reply]
Yes, I did not want to leave it blank, so I filled it with a one to see what the result would be if I had always bet on the home team.

**Reply by kyleskom** [reply]
Thats concerning, will take a look at it deeper and maybe start training some new models.

**Comment by trevon55** [comment]
This is an interesting project, and perhaps I have missed something, but appears you are not using player level data or models to make these predictions? I also cannot see rest, travel, consecutive home and road games, and some other important metadata being accounted for.

Edit: Reading further I can see that such queries have been made previously by others, and that time (as always) is a constraint. This is a great piece of work considering it started more as a learning exercise than anything else. However, I wouldn't recommend anybody bet on predictions made without accounting for a lot more then the moneyline model you have currently does, there is essentially no way you can have positive EV with the dataset currently in use.

**Reply by kyleskom** [reply]
I would agree. Treat this more as 1 tool in a tool belt as opposed to a single point of truth. Do hope to bring in more data you mentioned and player data down the road!

**Comment by JackPayne123** [comment]
I would love to see this done for O/U predictions maybe something could work with using the confidence percentage?

**Reply by ghost** [reply]
I tried it and the performance was worse than ML, even for only the predictions above 80%. I don't have the exact results available on the machine I'm currently using, but I will update them here later

**Reply by JackPayne123** [reply]
That's a shame, would love to see the sheet when you have a chance, thanks mate.

**Reply by ghost** [reply]
https://docs.google.com/spreadsheets/d/1HlFsI7zftKW1l4sDs1Xt08ty-nqFXgvI/edit?usp=sharing&ouid=103180396497360233850&rtpof=true&sd=true

edit: I did not spend more than three minutes looking the spreadsheet, so there is a high chance that something may be incorrect.

**Reply by JP23000** [reply]
So I have noticed that on the OU, the higher probabilities dont hit as often.  With that said over the last 3 months I am hitting 70% of bets on confidence level 48%-60% and 59% overall.  Which is amazing for 52.5% implied probability payout.

**Reply by JP23000** [reply]
Chalk it up to sports variance.  Models are built by humans and AI has the same fundamental flaws that we do to this point.  It looks at the all the same data as we do and it is inherent to the same bias.  Which is why you see it miss more at higher confidences, all the signs point heavily to one thing.... but sports variance takes over and on any given day something happens.  Spurs blew up the spread this year in games where they were given +10 points against the top teams in the league.  What it really excels at is finding the patterns in the games that are two close for us to decipher.

**Reply by ghost** [reply]
@JP23000 what is your sample size

**Reply by JP23000** [reply]
Sample size for the results on the OU?

only 150 games.  I didn't get onto this fully and tracking till mid march.

**Reply by JP23000** [reply]
I was 50/50 on OU's over 70%.  Also 3-0 on OU picks it kicks out that are under 50%.  It's hilarious.  It crunches the numbers and goes... okay.  There is a 52% chance it goes over... but fuck it.  I got a feeling.  We're taking the under.  Actually today is one.  It kicked out Wolves and Nuggs under 222.5 with a 48.9% confidence.  I expect to lose this one now since I have talked about it and now fate is going to slap me.  Plus going 100% on under 50% confidence picks is unsustainable.  I am also 5-1 over the last 2 days so I expect a correction back to the 59% average I have had overall.

Also as a fun note. Using the Kelly Criterion betting system and laddering my units up with the confidence level netted me a 110% return for the last 3 weeks of the season.

**Reply by nealmick** [reply]
share spreadsheet!

**Reply by ghost** [reply]
https://docs.google.com/spreadsheets/d/1LxPq1ZF3k6BreB8p50vESP-7BTZXVrFs6abU7qxaq2c/edit#gid=1616360111

found this on Reddit, it's not mine. someone shares the picks of the model every day,

**Reply by nealmick** [reply]
I talked to that guy, seems cool, should probably credit project in posts.

**Reply by kyleskom** [reply]
Interesting stats thanks for sharing. Ya I saw his Reddit posts as well. Some credit or just at least directing to my repo would be nice but all good.

**Reply by obtim** [reply]
> https://docs.google.com/spreadsheets/d/1LxPq1ZF3k6BreB8p50vESP-7BTZXVrFs6abU7qxaq2c/edit#gid=1616360111
> 
> found this on Reddit, it's not mine. someone shares the picks of the model every day,
Please, Share the link to his posts on reddit. Can't find it :(

**Reply by FutureTech22** [reply]
is there anything like this for the neural network model? or just the XG-boost model?

**Comment by drewhb** [comment]
@notfermat How/where did you get 74.9% model? I don't see any models with that percentage and after training I don't get any with that high a percentage. Maybe its just a matter of when you trained the data for the season since it was a few months ago?

**Reply by sonymag** [reply]
> @notfermat How/where did you get 74.9% model? I don't see any models with that percentage and after training I don't get any with that high a percentage. Maybe its just a matter of when you trained the data for the season since it was a few months ago?

I have 69.1% ML. Just keep training bro!

**Reply by rpassaro** [reply]
By this do you mean we just keep running the training commands? How do you know the %?

**Comment by AaronAshery** [comment]
> I backtested the "XGBoost_74.9%_ML-2.json" model on the 2022-23 season.
> 
> I downloaded data from teams using the `Get_Data` module, and then created a new function to acquire all historical odds from FanDuel using the `SbrOddsProvider` library. Next, I executed the `Create_Games` module to create the 2022-23 dataset.

Is this in the repo? if not would you mind sharing this function? I'd like to maybe test out how it would do betting based on EV and kelly

**Comment by MMSahin** [comment]
@lpUUzVRWtKgp can you please share the function in the repo or somehow else ? :) 

it would be great if you could do this in comprahension between NN and XGB Model. 

Big thanks!

**Comment by Hasan-Reza** [comment]
Can someone please explain how to get the Test_Current_season.py file running. It is trying to read from a dataset that does not exists in the repo nor is it being created.

**Comment by sonymag** [comment]
> I backtested the "XGBoost_74.9%_ML-2.json" model on the 2022-23 season.
> 
> I downloaded data from teams using the `Get_Data` module, and then created a new function to acquire all historical odds from FanDuel using the `SbrOddsProvider` library. Next, I executed the `Create_Games` module to create the 2022-23 dataset.
> 
> The 2022-23 dataset now contains two additional columns: `home_team_odds` and `away_team_odds`. These are necessary for the expected value calculation.
> 
> Once all steps were completed, I ran the `Test_Current_Season` script. Rather than returning only the winner prediction, I made a few modifications to the code so that it now returns the expected value, the probability, and the winner prediction. It only returns the team with the highest expected value.
> 
> For a $1 dollar per bet, the graph starts one month after the season start.
> 
> ![image](https://user-images.githubusercontent.com/43114210/215239389-e3437d2d-f258-4a24-bab7-32359d3615b6.png)
> 
> I utilized WPS Spreadsheets, and encountered a few difficulties while attempting to upload to Google Sheets and online Excel. I will provide the link to the .xlsx file, but here is a picture that displays all the information on the spreadsheet:
> 
> ![Current-Season-Test](https://user-images.githubusercontent.com/43114210/215239293-5d9ef0d2-260d-46c2-86be-ee45d6a3e19b.png)
> 
> Link to .xlsx file: https://mega.nz/file/YNgzxBwa#fo3Sp10MeefLl22z6qNAU5e2s7I4Q9MaOPd9s2ziAn8

Ty, now I have 74.8% for XGBoost Model ML and 62.2 for OU.

**Reply by mohcly** [reply]
the link is dead bro, please can I have a new one?

**Reply by rpassaro** [reply]
Could you share the 62.2% file please?
